core/integration/module_test.go

```diff
@@ -2088,222 +2485,437 @@ func (m *Test) Fn() (*Obj, error) {
-		require.Error(t, err)
-		require.ErrorContains(t, err, `Cannot query field "dep" on type "Query".`)
+		t.Run("direct", func(t *testing.T) {
+			t.Parallel()
+			_, err := ctr.
+				WithNewFile("src/main.py", dagger.ContainerWithNewFileOpts{
+					Contents: `import dagger
+
+@dagger.object_type
+class Obj:
+    foo: dagger.DepObj = dagger.field()
+
+@dagger.object_type
+class Test:
+    @dagger.function
+    def fn(self) -> Obj:
+        ...
+`,
+				}).
+				With(daggerFunctions()).
+				Stdout(ctx)
+			require.Error(t, err)
+			require.ErrorContains(t, err, fmt.Sprintf(
+				"object %q field %q cannot reference external type from dependency module %q",
+				"Obj", "foo", "dep",
+			))
+		})
+
+		t.Run("list", func(t *testing.T) {
+			t.Parallel()
+			_, err := ctr.
+				WithNewFile("src/main.py", dagger.ContainerWithNewFileOpts{
+					Contents: `import dagger
+
+@dagger.object_type
+class Obj:
+    foo: list[dagger.DepObj] = dagger.field()
+
+@dagger.object_type
+class Test:
+    @dagger.function
+    def fn(self) -> list[Obj]:
+        ...
+`,
+				}).
+				With(daggerFunctions()).
+				Stdout(ctx)
+			require.Error(t, err)
+			require.ErrorContains(t, err, fmt.Sprintf(
+				"object %q field %q cannot reference external type from dependency module %q",
+				"Obj", "foo", "dep",
+			))
+		})
 	})
 }
 
-func TestModuleCodegenonDepChange(t *testing.T) {
+var useGoInner = `package main
+
+type Dep struct{}
+
+func (m *Dep) Hello() string {
+	return "hello"
+}
+`
+
+var useGoOuter = `package main
+
+import "context"
+
+type Use struct{}
+
+func (m *Use) UseHello(ctx context.Context) (string, error) {
+	return dag.Dep().Hello(ctx)
+}
+`
+
+var usePythonInner = `from dagger import function
+
+@function
+def hello() -> str:
+    return "hello"
+`
+
+var usePythonOuter = `from dagger import dag, function
+
+@function
+def use_hello() -> str:
+    return dag.dep().hello()
+`
+
+func TestModuleUseLocal(t *testing.T) {
 	t.Parallel()
 
 	c, ctx := connect(t)
 
-	t.Run("go uses go", func(t *testing.T) {
-		t.Parallel()
+	type testCase struct {
+		sdk    string
+		source string
+	}
 
-		modGen := c.Container().From(golangImage).
-			WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
-			WithWorkdir("/work/dep").
-			With(daggerExec("mod", "init", "--name=dep", "--sdk=go")).
-			WithWorkdir("/work").
-			With(daggerExec("mod", "init", "--name=use", "--sdk=go")).
-			WithNewFile("/work/dep/main.go", dagger.ContainerWithNewFileOpts{
-				Contents: useInner,
-			}).
-			WithNewFile("/work/main.go", dagger.ContainerWithNewFileOpts{
-				Contents: useOuter,
-			}).
-			With(daggerExec("mod", "install", "./dep"))
+	inner := []testCase{
+		{
+			sdk:    "go",
+			source: useGoInner,
+		},
+		{
+			sdk:    "python",
+			source: usePythonInner,
+		},
+	}
 
-		logGen(ctx, t, modGen.Directory("."))
+	outer := []testCase{
+		{
+			sdk:    "go",
+			source: useGoOuter,
+		},
+		{
+			sdk:    "python",
+			source: usePythonOuter,
+		},
+	}
```

Comment:
I'm not sure if there's value in these permutations. We can also exclude `dep.sdk == use.sdk` or just use "go" as _inner_ and iterate on _outer_.